File: vue/src/features/events/types/events.ts
import type { ApiQueryValue } from '@/shared/types/common'

export interface SecurityEventItem {
  id: number
  layer: string
  provider: string | null
  provider_event_id: string | null
  provider_site_id: string | null
  provider_site_name: string | null
  provider_site_domain: string | null
  action: string
  reason: string
  details_json: string | null
  source_ip: string
  dest_ip: string
  source_port: number
  dest_port: number
  protocol: string
  http_method: string | null
  uri: string | null
  http_version: string | null
  created_at: number
  handled: boolean
  handled_at: number | null
}

export interface SecurityEventsResponse {
  total: number
  limit: number
  offset: number
  events: SecurityEventItem[]
}

export interface BehaviorProfileItem {
  identity: string
  source_ip: string | null
  latest_seen_at: number
  score: number
  dominant_route: string | null
  focused_document_route: string | null
  focused_api_route: string | null
  distinct_routes: number
  repeated_ratio: number
  document_repeated_ratio: number
  api_repeated_ratio: number
  interval_jitter_ms: number | null
  document_requests: number
  api_requests: number
  non_document_requests: number
  challenge_count_window: number
  session_span_secs: number
  flags: string[]
  latest_route: string
  latest_kind: string
  blocked: boolean
  blocked_at: number | null
  blocked_expires_at: number | null
  blocked_reason: string | null
}

export interface BehaviorProfilesResponse {
  total: number
  profiles: BehaviorProfileItem[]
}

export interface FingerprintProfileItem {
  identity: string
  identity_kind: string
  source_ip: string | null
  first_seen_at: number
  last_seen_at: number
  first_site_domain: string | null
  last_site_domain: string | null
  first_user_agent: string | null
  last_user_agent: string | null
  total_security_events: number
  total_behavior_events: number
  total_challenges: number
  total_blocks: number
  latest_score: number | null
  max_score: number
  latest_action: string | null
  reputation_score: number
  notes: string
}

export interface FingerprintProfilesResponse {
  total: number
  profiles: FingerprintProfileItem[]
}

export interface BehaviorSessionItem {
  session_key: string
  identity: string
  source_ip: string | null
  site_domain: string | null
  opened_at: number
  last_seen_at: number
  event_count: number
  challenge_count: number
  block_count: number
  latest_action: string | null
  latest_uri: string | null
  latest_reason: string | null
  dominant_route: string | null
  focused_document_route: string | null
  focused_api_route: string | null
  distinct_routes: number
  repeated_ratio: number
  document_repeated_ratio: number
  api_repeated_ratio: number
  document_requests: number
  api_requests: number
  non_document_requests: number
  interval_jitter_ms: number | null
  session_span_secs: number
  flags: string[]
}

export interface BehaviorSessionsResponse {
  total: number
  sessions: BehaviorSessionItem[]
}

export interface EventsQuery extends Record<string, ApiQueryValue> {
  limit?: number
  offset?: number
  layer?: string
  provider?: string
  provider_site_id?: string
  source_ip?: string
  action?: string
  blocked_only?: boolean
  handled_only?: boolean
  sort_by?: string
  sort_direction?: 'asc' | 'desc'
}

export interface BlockedIpItem {
  id: number
  provider: string | null
  provider_remote_id: string | null
  ip: string
  reason: string
  blocked_at: number
  expires_at: number
}

export interface BlockedIpsResponse {
  total: number
  limit: number
  offset: number
  blocked_ips: BlockedIpItem[]
}

export interface BlockedIpsQuery extends Record<string, ApiQueryValue> {
  limit?: number
  offset?: number
  source_scope?: 'all' | 'local' | 'remote'
  provider?: string
  ip?: string
  keyword?: string
  active_only?: boolean
  blocked_from?: number
  blocked_to?: number
  sort_by?: string
  sort_direction?: 'asc' | 'desc'
}

export interface BlockedIpCreatePayload {
  ip: string
  reason: string
  duration_secs?: number
}

export interface BlockedIpsBatchUnblockPayload {
  ids: number[]
}

export interface BlockedIpsBatchUnblockResponse {
  success: boolean
  requested: number
  unblocked: number
  failed: number
  failed_ids: number[]
  message: string
}

export interface BlockedIpsCleanupExpiredPayload {
  source_scope?: 'all' | 'local' | 'remote'
  provider?: string
  blocked_from?: number
  blocked_to?: number
  expires_before?: number
}

export interface BlockedIpsCleanupExpiredResponse {
  success: boolean
  cleaned: number
  runtime_unblocked: number
  message: string
}

File: src/storage/models/events.rs
use super::super::unix_timestamp;

#[cfg_attr(not(feature = "api"), allow(dead_code))]
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct SecurityEventEntry {
    pub id: i64,
    pub layer: String,
    pub provider: Option<String>,
    pub provider_event_id: Option<String>,
    pub provider_site_id: Option<String>,
    pub provider_site_name: Option<String>,
    pub provider_site_domain: Option<String>,
    pub action: String,
    pub reason: String,
    pub details_json: Option<String>,
    pub source_ip: String,
    pub dest_ip: String,
    pub source_port: i64,
    pub dest_port: i64,
    pub protocol: String,
    pub http_method: Option<String>,
    pub uri: Option<String>,
    pub http_version: Option<String>,
    pub created_at: i64,
    pub handled: bool,
    pub handled_at: Option<i64>,
}

#[derive(Debug, Clone, sqlx::FromRow)]
pub struct BotIpCacheEntry {
    pub provider: String,
    pub ranges_json: String,
    pub last_refresh_at: Option<i64>,
    pub last_success_at: Option<i64>,
    pub last_error: Option<String>,
    pub updated_at: i64,
}

#[cfg_attr(not(feature = "api"), allow(dead_code))]
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct BlockedIpEntry {
    pub id: i64,
    pub provider: Option<String>,
    pub provider_remote_id: Option<String>,
    pub ip: String,
    pub reason: String,
    pub blocked_at: i64,
    pub expires_at: i64,
}

#[cfg_attr(not(feature = "api"), allow(dead_code))]
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct FingerprintProfileEntry {
    pub identity: String,
    pub identity_kind: String,
    pub source_ip: Option<String>,
    pub first_seen_at: i64,
    pub last_seen_at: i64,
    pub first_site_domain: Option<String>,
    pub last_site_domain: Option<String>,
    pub first_user_agent: Option<String>,
    pub last_user_agent: Option<String>,
    pub total_security_events: i64,
    pub total_behavior_events: i64,
    pub total_challenges: i64,
    pub total_blocks: i64,
    pub latest_score: Option<i64>,
    pub max_score: i64,
    pub latest_action: Option<String>,
    pub reputation_score: i64,
    pub notes: String,
}

#[cfg_attr(not(feature = "api"), allow(dead_code))]
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct BehaviorSessionEntry {
    pub session_key: String,
    pub identity: String,
    pub source_ip: Option<String>,
    pub site_domain: Option<String>,
    pub opened_at: i64,
    pub last_seen_at: i64,
    pub event_count: i64,
    pub challenge_count: i64,
    pub block_count: i64,
    pub latest_action: Option<String>,
    pub latest_uri: Option<String>,
    pub latest_reason: Option<String>,
    pub dominant_route: Option<String>,
    pub focused_document_route: Option<String>,
    pub focused_api_route: Option<String>,
    pub distinct_routes: i64,
    pub repeated_ratio: i64,
    pub document_repeated_ratio: i64,
    pub api_repeated_ratio: i64,
    pub document_requests: i64,
    pub api_requests: i64,
    pub non_document_requests: i64,
    pub interval_jitter_ms: Option<i64>,
    pub session_span_secs: i64,
    pub flags_json: String,
}

#[cfg_attr(not(feature = "api"), allow(dead_code))]
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct BehaviorEventEntry {
    pub id: i64,
    pub security_event_id: Option<i64>,
    pub identity: String,
    pub session_key: String,
    pub source_ip: String,
    pub site_domain: Option<String>,
    pub http_method: Option<String>,
    pub uri: Option<String>,
    pub action: Option<String>,
    pub reason: String,
    pub score: i64,
    pub dominant_route: Option<String>,
    pub focused_document_route: Option<String>,
    pub focused_api_route: Option<String>,
    pub distinct_routes: i64,
    pub repeated_ratio: i64,
    pub document_repeated_ratio: i64,
    pub api_repeated_ratio: i64,
    pub document_requests: i64,
    pub api_requests: i64,
    pub non_document_requests: i64,
    pub interval_jitter_ms: Option<i64>,
    pub challenge_count_window: i64,
    pub session_span_secs: i64,
    pub flags_json: String,
    pub created_at: i64,
}

#[derive(Debug, Clone)]
pub struct SecurityEventRecord {
    pub layer: String,
    pub provider: Option<String>,
    pub provider_event_id: Option<String>,
    pub provider_site_id: Option<String>,
    pub provider_site_name: Option<String>,
    pub provider_site_domain: Option<String>,
    pub action: String,
    pub reason: String,
    pub details_json: Option<String>,
    pub source_ip: String,
    pub dest_ip: String,
    pub source_port: i64,
    pub dest_port: i64,
    pub protocol: String,
    pub http_method: Option<String>,
    pub uri: Option<String>,
    pub http_version: Option<String>,
    pub created_at: i64,
    pub handled: bool,
    pub handled_at: Option<i64>,
}

#[derive(Debug, Clone)]
pub struct BlockedIpRecord {
    pub provider: Option<String>,
    pub provider_remote_id: Option<String>,
    pub ip: String,
    pub reason: String,
    pub blocked_at: i64,
    pub expires_at: i64,
}

impl SecurityEventRecord {
    pub fn now(
        layer: impl Into<String>,
        action: impl Into<String>,
        reason: impl Into<String>,
        source_ip: impl Into<String>,
        dest_ip: impl Into<String>,
        source_port: u16,
        dest_port: u16,
        protocol: impl Into<String>,
    ) -> Self {
        Self {
            layer: layer.into(),
            provider: None,
            provider_event_id: None,
            provider_site_id: None,
            provider_site_name: None,
            provider_site_domain: None,
            action: action.into(),
            reason: reason.into(),
            details_json: None,
            source_ip: source_ip.into(),
            dest_ip: dest_ip.into(),
            source_port: i64::from(source_port),
            dest_port: i64::from(dest_port),
            protocol: protocol.into(),
            http_method: None,
            uri: None,
            http_version: None,
            created_at: unix_timestamp(),
            handled: false,
            handled_at: None,
        }
    }
}

impl BlockedIpRecord {
    pub fn new(
        ip: impl Into<String>,
        reason: impl Into<String>,
        blocked_at: i64,
        expires_at: i64,
    ) -> Self {
        Self {
            provider: None,
            provider_remote_id: None,
            ip: ip.into(),
            reason: reason.into(),
            blocked_at,
            expires_at,
        }
    }
}
